feat: add ADR for certs management (#83)
bf2fc6cc711aee1a0c2a/architecture: _adr/89/index.adoc

---
num: 89 
title: A bf2 subdomain & wildcard certificate per data plane cluster 
status: "Draft" # One of Draft, Accepted, Rejected
authors:
  - "Manyanda Chitimbo"
  - "Steven Hawkins"
  - "Peter Braun"
tags:
  - "kafka" # e.g. kafka, connectors, registry
applies_padrs: # What PADRs does this ADR apply?
applies_patterns: # What APs does this ADR apply?
---

// Top style tips:
// \* Use one sentence per line
// \* No unexpanded acronyms
// \* No undefined jargon

// No need for a title heading, it's added by the template

## Context and Problem Statement
// What is the background against which this decision is being taken?

> The ADR intends to solve the same problem as the one described in https://architecture.appservices.tech/adr/5/[ADR-5: Kafka Cluster Certificate Generation].

Author: It is briefly stated in the goals. I can additionally copy & paste the problem as state in the ADR 5, clearly stating that the issue at the moment is that the process is manual and it exposes shared the certs to customer data planes. What do you think?

> The ADR intends to solve the same problem as the one described in https://architecture.appservices.tech/adr/5/[ADR-5: Kafka Cluster Certificate Generation].

Author: took the liberty of writing down what I propose above. I'll push it for second round of review once you've the time.

## Goals
// Bulleted list of outcomes that this ADR, if accepted, should help achieve
- Automated certificate management (generation, renewal, revocation) per data plane cluster
- Avoid sharing same certificate between enterprise data plane clusters
- Avoid exposing domain registrar & certificate issuers credentials onto enterprise data plane clusters

## Non-goals
// Bulleted list of outcomes that this ADR is not trying to achieve.
- Automated certificate management per kafka instance.
- Creating a pattern for and/or developing a central certificate management service

## Current situation
// Where are we now?

The https://architecture.appservices.tech/adr/5/[ADR-5: Kafka Cluster Certificate Generation] describes the current situation where one certificate is shared across data planes. The certificate is renewed manually using the https://github.com/bf2fc6cc711aee1a0c2a/kas-sre-sops/blob/main/sops/key_management/control_plane/dataplane_certificate.asciidoc[data plane certificate SOP]

## Proposal
// What is the decision being proposed

> The proposed solution is going to use https://Letsencrypt.org[Letsrncrypt] to automatically manage certificates.

Reviewer: ```suggestion
The proposed solution is going to use https://Letsencrypt.org[Letsencrypt] to automatically manage certificates.
```

The solution consists of: 

1. Generating one wildcard certificate per data plane cluster.
For each new data plane, there will be a wildcard certificate for the  the `*.<cluster-id>.kafka.bf2.dev` subdomain where `<cluster-id>` is unique for each cluster.

> 2. The fleet manager in its reconcilers, will use https://github.com/caddyserver/certmagic[certmagic library] to generate certificate for each data plane the certificate and store them in a secure Storage. 

Reviewer: ```suggestion
2. The fleet manager in its reconcilers will use https://github.com/caddyserver/certmagic[certmagic library] to generate certificate for each data plane and store them in a secure Storage. 
```

> When a kafka is assigned to a given data plane, all routes created will be under the `*.<cluster-id>.kafka.bf2.dev` domain. 

Author: No, today all the routes for Kafka are a subdomain of `kafka.bf2.dev`. It'll only be the case going forward with this ADR is implemented. This is one of the reason why performing a transparent migration is hard.

For example, if the cluster id is `mycluster` then Kafka boostrap hosts will have the form of `foobar.mycluster.kafka.bf2.dev`.
During Kafka reconciliation between the Fleetshard and the Fleet manager, certificates will be fetched from the secure storage by the fleet manager and they'll be passed down to the data plane via the existing certificate passing mechanism i.e via the ManagedKafka CR. 
3. No migration to be performed for existing data plane.
The proposed solution will be applied to only new data plane clusters.  

> Existing data plane clusters will continue to share the wildcard certificate on the `*.kafka.bf2.dev` domain.  

Author: Sounds good. I am also open to suggestions on whether & how we could consider migrating the existing data planes to using this solution. My suggestion was that we make this change a non disruptive one to existing users / Kafkas but I am open to alternative suggestion e.g go ahead with the change all over and offer a migration guide

> Existing data plane clusters will continue to share the wildcard certificate on the `*.kafka.bf2.dev` domain.  

Reviewer 2: If we change the existing certificate, would that be disruptive? From what I understand it would force client to reconnect to the brokers, but other than that, are there any side effects?

> Existing data plane clusters will continue to share the wildcard certificate on the `*.kafka.bf2.dev` domain.  

Reviewer 3: I inferred that we would need to change the Kafka URL because the Kafka broker URL is simething like `kafka-uuid-broker1.kafka.bf2.org`. We did not anticipate the need.

> Existing data plane clusters will continue to share the wildcard certificate on the `*.kafka.bf2.dev` domain.  

Reviewer 3: meaning the need for separate subdomains per kafka deployments

> Existing data plane clusters will continue to share the wildcard certificate on the `*.kafka.bf2.dev` domain.  

Author: > I inferred that we would need to change the Kafka URL because the Kafka broker URL is simething like kafka-uuid-broker1.kafka.bf2.org. We did not anticipate the need.

That's the issue. 

> ah but if that's the case, why not ask for a multi named cert?

I've not investigate how this will look like but from face value it looks like we'll need to generate a certificate per kafka instance. This is not what's advertised here in the ADR, but this would solve the issue, however there are challenges that are related to this:
1. We'll like need to ask for overrrides of letsencrypt limits as generating per Kafka we are more likely to hit them often
2. In the near by future, we'll have Kafka expansion feature. Which may add some complexity to the certificate management as new brokers are scaled up / brokers are scaled down we'll need to make sure that the certificate knows of new brokers or doesn't know old brokers.

The shared certificate for the already existing data planes will now be automatically renewed by the fleet manager.

> 4. Certificate renewals will also be handled automatically by the fleet manager using the https://github.com/caddyserver/certmagic[certmagic library]. 

Author: +1 makes sense. 

in the long run, we hope there will be a central service for ingress management e.g Hybrid Cloud Gateway

> 4. Certificate renewals will also be handled automatically by the fleet manager using the https://github.com/caddyserver/certmagic[certmagic library]. 

Reviewer 2: @emmanuelbernard do you still think it should be a library, even though at some point certificates will be handled by HCG as presented by Craig? We're already using a library to implement this (certmagic).

> 4. Certificate renewals will also be handled automatically by the fleet manager using the https://github.com/caddyserver/certmagic[certmagic library]. 

Reviewer 3: I'd need to see the certmagic API level honestly. We can discuss it.

> 4. Certificate renewals will also be handled automatically by the fleet manager using the https://github.com/caddyserver/certmagic[certmagic library]. 

Author: The default certmagic API is pretty easy to work with. But we'll be building things on top of it and mostly the Vault Storage which is not offered by default. So these things that we build on top of cert magic can as well as be extracted as a library/sdk.

> 4. Certificate renewals will also be handle automatically by the fleet manager using the https://github.com/caddyserver/certmagic[certmagic library]. 

Reviewer: ```suggestion
4. Certificate renewals will also be handled automatically by the fleet manager using the https://github.com/caddyserver/certmagic[certmagic library]. 
```

If this is on a enterprise data plane cluster without good security standards/policy, then anyone with the private key can impersonate the server or eavesdrop on a connection with the server. 
For non enterprise data plane clusters, this is mitigated by only allowing SREs to have the priviledge / permissions to see these secrets.
For enterprise data plane clusters, customers will be advised to do so as well.

Morever, since there will be a different certificate per data plane clusters, if one certificate is compromised, it won't affect others and it can be safely revoked. 
The proposed solution also limits the possibility of an owner of a enterprise data plane cluster, eavesdropping on another data plane cluster that they do not own as the certificate keys will be different.

## Alternatives Considered / Rejected

1. Deploying https://www.redhat.com/sysadmin/cert-manager-operator-openshift[cert-manager operator] to the data plane. Rejected to avoid exposing domain registrar credentials and to avoid adding another component that consumes data plane resources and that needs to be monitored.

> 2. Migrating existing data plane (and their Kafkas). This change is disruptive to existing customers hence why it was rejected. 

Reviewer 3: I understand but then let's put in the threat model that this  wildcard certificate of a lower level will allow impersonation of the enterprise kafka cluster.

The Kafka fleet manager will thus ensure that existing Kafkas on existing data planes clusters continue to operate using a shared wildcard certificate on the `*.kafka.bf2.dev` domain i.e the proposed solution will only be applied to new data plane clusters.

3. Deploying https://www.redhat.com/sysadmin/cert-manager-operator-openshift[cert-manager operator] in the Control Plane. 

> The Control plane then handles the pushing the Certificate reqeuest CR, watching for the created secrets and storing them securely. 

Reviewer: ```suggestion
The Control plane then handles the pushing of the Certificate request CR, watching for the created secrets and storing them securely. 
```
